content/posts/2022-05-13-calculating-type-sets.markdown

```diff
@@ -0,0 +1,511 @@
+at all objects for which `f` is `true`:
+
+    S[f] := { x∈U | f(x) }
+
+We use the index form `S[f]` to indicate that `S` depends on the function `f`.
+
+Likewise, if we have any set `S⊆U`, we can create a boolean function from it:
+
+    f[S]: U → B
+          x ↦ x∈S
+
+That is, the function returns `true` if an element is in `S` and `false`
+otherwise.
+
+An interesting aspect of this, is that we can translate between boolean
+operators and set operators this way. For example, set union `∪` becomes
+equivalent to logical disjunction `∨`:
+
+    S[f] ∪ S[g] = { x∈U | f(x) } ∪ { x∈U | g(x) }
+                = { x∈U | f(x) ∨ g(x) }
+
+    f[S](x) ∨ f[T](x) = x∈S ∨ x∈T = x∈(S∪T)
+
+Similarly, `∩` becomes equivalent to `∧`. `¬` becomes equivalent to set
+complement: `S[¬f] = { x∈U | ¬f(x) } = { x∈U | x∉S }`, which we can write as
+`U\S`.
+
+The boolean formulas we looked at in the previous section are a form of these
+boolean functions. They take some boolean arguments and return a boolean, e.g.
+
+      f: B³ → B
+    (X,Y,Z) ↦ ((¬X∧Y)∨Z)∧(X∨¬Y)
+
+[It turns out](https://en.wikipedia.org/wiki/Functional_completeness) that *every*
+boolean function with boolean parameters can be written as a logical formula -
+in particular, every such function has a CNF and a DNF.
+
+If we want to do that for other sets (apart from `B^n`, that is), we first have
+to be able to translate it into a set of boolean variables. As we will see,
+this can be easily done in the case we are interested in.
+
+All of this allows us to talk about what we mean by “computing a set is hard”.
+With this equivalency, checking if a given object is in a set becomes the
+*decision* problem of SAT - deciding if `x∈S` is equivalent to deciding if
+`f[S](x)`evaluates to `true`. And checking if a set is empty becomes the
+*solution* problem of SAT, as it is equivalent to verifying that `f[S]` is not
+satisfiable.
+
+With this, let us look at the specific sets we are interested in.
+
+## Basic interfaces as type sets
+
+Interfaces in Go have always described sets of types. For example, the interface
+
+```go
+type S interface {
+    X()
+    Y()
+    Z()
+}
+```
+
+is “the set of all types which have a method `X()` and a method `Y()` and a method `Z()`”.
+
+We also have a way to express set intersection, using [interface embedding](https://go.dev/ref/spec#Embedded_interfaces):
+
+
+```go
+type S interface { X() }
+type T interface { Y() }
+type U interface {
+    S
+    T
+}
+```
+
+This expresses `T=S∩T` as an interface. We can also view the predicate “has a
+method `X()`” as a boolean variable and think of this as the formula `X∧Y`.
+
+Surprisingly, there is also a simple form of negation. It happens implicitly,
+because a type can not have two methods with the same name, but different
+types. So, implicitly, if a type has a method `X()`, it does *not* have a
+method `X() int`, for example[^5]:
+
+```go
+type X interface { X() }
+type NotX interface{ X() int }
```

Comment:
There are, of course, infinitely many other interfaces that could also be `NotX`.

Reply:
Do you think this needs another mention? In my defense, I do say "for example".

Reply:
No, I suppose not.